File: src/cpp/src/subscriber.cpp
#include "Subscriber.hpp"

#include <string>

Subscriber::Subscriber() : context(1), socket(context, zmq::socket_type::pull)
{
    socket.connect("tcp://localhost:5555"); // Connects to Python Publisher
}

Subscriber::~Subscriber()
{
    socket.close();
}

void Subscriber::subscribe(const std::function<bool(const AudioPacket &)> &callback)
{
    while (true)
    {
        AudioPacket packet = receive();
        if (!callback(packet))
        {
            break;
        }
    }
}

AudioPacket Subscriber::receive()
{
    zmq::message_t zmq_message;
    auto result = socket.recv(zmq_message, zmq::recv_flags::none);

    if (!result)
    {
        throw std::runtime_error("Failed to receive message from socket");
    }

    auto buffer = static_cast<uint8_t *>(zmq_message.data());
    auto data = std::vector(buffer, buffer + zmq_message.size());
    return AudioPacket::deserialize(data);
}

File: src/zmq_experiment/sub/subscriber.py
from typing import AsyncGenerator

import zmq
import zmq.asyncio

from ..audio_packet import AudioPacket


class Subscriber:
    _context: zmq.asyncio.Context
    _socket: zmq.asyncio.Socket

    def __init__(self):
        self._context = zmq.asyncio.Context()
        self._socket = self._context.socket(zmq.PULL)
        self._socket.connect("tcp://localhost:5555")

    async def subscribe(self) -> AsyncGenerator[AudioPacket, None]:
        while True:
            data = await self._socket.recv()
            yield AudioPacket.deserialize(data)
